implementation ChromBackend

philouail opened the pull request:
So there is still a lot lot to do but as discussed I am trying to reduce the size of the PR. I hope that does not make it too confusing.

The big changes are the clear separation between what is a peaksVariables and what is a chromvariables. 
See `ChromBackend-functions.R`

Also implementation of respective function for each such as `validPeaksData()` and `validChromData()`

Quick question:
- I added a replacement method for chromIndex, is that okay or in general we don't want people to change that ?
- is it a problem that `chromData()` returns a `data.frame` and not `DataFrame` ? I don't really see the issue for now..

Next I will add the `ChromBackendMemory()` implementation, i have the code, just need to clean it.
Also will look into some maintainer stuff, if you have advices don't hesitate.

philouail commented:
I changed the .yml file, the checks should not fail on warnings anymore :)

jorainer commented on R/ChromBackend.R:
```diff
@@ -134,28 +134,24 @@
 #'
 #' - `chromData()`, `chromData<-`: gets or sets general chromatogram metadata
-#'   (annotation).  `chromData()` returns a `DataFrame`, `chromData<-` expects
+#'   (annotation). `chromData()` returns a `DataFrame`, `chromData<-` expects
```
I would suggest you also describe the `drop` parameter here to clearly define what `drop = TRUE` and `drop = FALSE` is expected to do. Same also for the `peaksData()` documentation below. And you also need to have a `@param drop` documentation entry (maybe you have it already but I did not see it here). We need to be very clear with this parameter, especially because we don't have that in *Spectra*.

jorainer reviewed (changes requested):
Looks good Phili! Thanks! Just some minor change requests.

Regarding the question if `chromData()` should return a `DataFrame` or `data.frame`, I don't see a problem for `data.frame` at the moment - so, given that it is also slightly more efficient, I would maybe change to `data.frame` instead of `DataFrame`.

philouail commented on R/ChromBackend-functions.R:
```diff
@@ -1,43 +1,17 @@
-#' @importFrom MsCoreUtils common
-#'
-#' @author Sebastian Gibb, Johannes Rainer
-.values_match_mz <- function(x, mz, ppm = 20, tolerance = 0) {
```
This is used later on for the filter function that you implemented. I kept it on another file locally, but I only wanted things related to what is implemented now to be in this PR. So it will probably come back when I start implementing filter functions :)

jorainer replied:
Thanks for the clarification!

philouail commented on R/AllGenerics.R:
```diff
@@ -48,3 +51,6 @@ setGeneric("productMzMin<-", function(object, value)
 setGeneric("selectChromVariables", function(object, ...)
     standardGeneric("selectChromVariables"))
+#' @rdname ChromBackend
+setGeneric("reset", function(object, ...)
```
mmm not that I can see ! It's also in the `AllGenerics.R` in `Spectra` :)

jorainer replied:
OK - we might then need (later at some point) add that to *ProtGenerics*.

philouail commented on R/ChromBackend-functions.R:
```diff
@@ -124,17 +97,46 @@ validChromData <- function(x = data.frame(), error = TRUE) {
+    if (anyNA(x))
+        return("'NA' values in dataStorage are not allowed.")
+    NULL
+}  ## what's this ?
```
mmm follow up question then, should the setValidity() be fully removed for `ChromBackend` ? And we make sure we properly use  validChromData() and validPeaksData() when necessary ?

jorainer replied:
Good question. I would be OK with removing it for now and then, if we see we need it, add it back. You, as package maintainer, can decide if you want to remove or not ;)

philouail commented:
Hi so I answered / addressed the comments, thanks for the advices ! I went through the documentation a bit more realised I missed removing some stuff also.
 tell me if there is something else :) and for moving the `reset()` to `ProtGenerics` also

jorainer commented on R/ChromBackend.R:
```diff
@@ -126,19 +126,20 @@
 #'   replacement method `chromData<-` (unless some internal caching mechanism
-#'   could be used).
+#'   could be used). `chromData()` should be implemented with the parameter
+#'   `drop` set to `FALSE` as default. this mean that by default the method
```
Please reformulate here *... as default. this mean...* - maybe with * ... as default. With `drop = FALSE` the method should return ...*.

jorainer commented on R/ChromBackend.R:
```diff
@@ -147,7 +148,11 @@
 #'   variables should be returned in each array. As default (minimum) columns
-#'   `"rtime"` and `"intensity"` have to be provided.
+#'   `"rtime"` and `"intensity"` have to be provided. `peaksData()` should be
+#'   implemented with the parameter `drop` set to `FALSE` as default. this mean
```
See previous comment.

jorainer reviewed (approved):
Looks good now, thanks Phili!

Before merging make sure you fix the two comments/issues in the documentation.

Regarding `reset()` - yes, we might want to add that to *ProtGenerics*, but for now define it also here. We can change later.
